### ISSUE TITLE
Why everything get's re rendered?

### ISSUE THREAD

**Issue by arvydasg** [bastibe/org-static-blog#126]
Hello!

Thank you for this great package, it has helped me to build my blog!

--------------------------------------------------
About the issue I am having:

We can see that only "hello" was added to the .org file:

![image](https://github.com/bastibe/org-static-blog/assets/78803192/e4cc60e1-9477-4389-ac76-0905bc3ecf3a)

When I execute my shell script named `build.sh` that runs this command:

```emacs -Q --script build-site.el```

build-site.el content:

```;; Set the package installation directory so that packages aren't stored in the
;; ~/.emacs.d/elpa path.
(require 'package)
(setq package-user-dir (expand-file-name "./.packages"))
(setq package-archives '(("melpa" . "https://melpa.org/packages/")
			 ("elpa" . "https://elpa.gnu.org/packages/")))

;; Initialize the package system
(package-initialize)
(unless package-archive-contents
  (package-refresh-contents))

(message "Package refresh complete!")

;; Install dependencies inside the ./packages directory
;; should install ONLY org-static-blog
(package-install 'org-static-blog)
(message "Package installation complete!")

;; stop making ~ files
(setq make-backup-files nil)

;; directories and urls
(setq org-static-blog-publish-title "yes.github.io")
(setq org-static-blog-publish-url "https://yes.github.io/private/html")
(setq org-static-blog-publish-directory "~/GIT/yes/html")
(setq org-static-blog-posts-directory "~/GIT/yes/org")
(setq org-static-blog-drafts-directory "~/GIT/yes/drafts")
(setq org-static-blog-index-length 5)
(setq org-static-blog-preview-date-first-p nil)
(setq org-static-blog-use-preview t)
(setq org-static-blog-enable-tags t)
(setq org-static-blog-no-post-tag "NONPOST")
(setq org-static-blog-enable-deprecation-warning nil)
;; (setq org-export-with-toc nil)            ;can add in individual file with #+OPTIONS: toc:1/nil
;; (setq org-export-with-section-numbers nil) ;can add in individual file with
;; #+OPTIONS: num:nil
(setq org-static-blog-page-header
"<meta name=\"author\" content=\"yes\">
<meta name=\"referrer\" content=\"no-referrer\">
<meta name=\"viewport\" content=\"initial-scale=1,width=device-width,minimum-scale=1\">
<link rel=\"stylesheet\" href=\"../static/style.css\"/>
<link rel=\"icon\" href=\"../static/ag.ico\">
<script src=\"../static/lightbox.js\"></script>
<script src=\"static/auto-render.min.js\"></script>")

(setq org-static-blog-page-preamble "<br>preamble here yo</br>")
(setq org-static-blog-page-postamble "<br>postamble in de house</br>")

(org-static-blog-publish t)

(message "Build complete!")
```

The html files are being generated, all good, BUT for some reason EVERYTHING get's rebuild. In my case all 41 files. Only the id tags change.. I don't really understand why. Is it intentional? I don't really like that it changes the "file modified" data, not sure if it translates to the date shown in the blog itself? Also all these git commits each time I make a change will make it unworkable to version control.

![image](https://github.com/bastibe/org-static-blog/assets/78803192/6e525c8e-363a-4344-b5d6-1f8fe0d79971)

**Comment by arvydasg** [comment]
I see this in the readme.org:

![image](https://github.com/bastibe/org-static-blog/assets/78803192/79ab6b4f-1073-4a10-8c74-54ae3f630bbe)

But even if the html files already exist, they are being re rendered.

**Comment by bastibe** [comment]
Only the changed org-file should be re-exported. Regrettably, this does change all the ids in the HTML file, as they seem to be randomly generated by org-mode on every export. (If anyone knows a way of preventing that, I'd be most interested!) Existing HTML files should not be changed, however.

You can check out [my own config](https://github.com/bastibe/.emacs.d/blob/master/init.el#L753) to see if there are any notable differences between yours and mine. You could also try to configure your org-static-blog with absolute directory paths instead of relative ones. I've run into no ends of problems with that sort of thing, so that's perhaps worth a try.

**Comment by arvydasg** [comment]
"Only the changed org-file should be re-exported." 
"Existing HTML files should not be changed, however."

I tried making a little experiment. Today is Dec 18th. I have launched emacs with "emacs -q" not to load any additional configuration.

![image](https://github.com/bastibe/org-static-blog/assets/78803192/7343224c-d101-40c5-80c2-dfe016219b45)

As we can see in the image I have 3 org files in /home/nixos/GIT/org-static-blog/org directory

- Dec 16 18:46 alpine.org
- Dec 11 19:24 crontab__code_linux.org
- Dec 16 10:01 windows-keybindings__windows.org

I have just generated them all, so we can see the html file creation time is Dec 18th, 06:23:

- Dec 18 06:23 alpine.html
- Dec 18 06:23 archive.html
- Dec 18 06:23 crontab__code_linux.html
- Dec 18 06:23 index.html
- Dec 18 06:23 rss.xml
- Dec 18 06:23 tag-code.html
- Dec 18 06:23 tag-linux.html
- Dec 18 06:23 tags.html
- Dec 18 06:23 tag-windows.html
- Dec 18 06:23 windows-keybindings__windows.html

I will go and modify one org file called "crontab__code_linux.org", which was last modified on Dec 11th.

![image](https://github.com/bastibe/org-static-blog/assets/78803192/04cc4c6e-27c0-461e-80e8-a3a8ab40540f)

Now I will regenerate the html files wiht my build command:

![image](https://github.com/bastibe/org-static-blog/assets/78803192/76363b42-6559-4843-af05-0393d7667170)

We can see that 6:44 I have changed the org file, then 06:45 I have regenerated the the html files and instead of only regenerating "crontab__code_linux.html", ALL html files got regenerated

![image](https://github.com/bastibe/org-static-blog/assets/78803192/46a63756-0f8a-4830-8d41-da88ff8c0474)

So in the "Archive" page or in the post page, the date that we will see will not be from

- #+date:       [2023-11-25 Sat 19:05] field that we specify in each org file (which would be my goal)
- from the actual org file creation date

But it wil always be the date of when the file was generated. Which will always be the date when you make a change to ANY of your .org file.

I wish there was a way to use these in org-static-blog export:

![image](https://github.com/bastibe/org-static-blog/assets/78803192/c458fcff-67df-40a2-9678-c09656de7281)

![image](https://github.com/bastibe/org-static-blog/assets/78803192/90b60f89-2a9d-43ac-97d7-e24a2eec4a3a)

So I would clearly see the date that I specify in org file, the a clear date when the file was created and when the file was modified.

In org-static-blog it seems that I only can see the date modifed, which changes with each change in any of the org file it seems.

**Comment by arvydasg** [comment]
Placed the above for reference, what I will try now is:
- create new posts with `org-static-blog-create-new-post` instead any other way
- try to use absolute directory paths instead of relative ones

will let you know how it goes

**Comment by arvydasg** [comment]
Okay so seems like having org-static-blog configuration in a separate file like I have(`build-site.el`) and then running it with `build.sh` script (like I described in my initial message) to generate the html files is responsible for this strange behavior where everything gets re-rendered.

IF I place my org-static-blog configuration(`build-site.el`) content into my emacs config, together with the rest of the packages - the behavior is proper, ONLY the modified .org files get re-rendered.

So basically this way of having multiple org-static-blog configs for different blogs does not really work(by the way, got this idea here - https://systemcrafters.net/publishing-websites-with-org-mode/building-the-site/ ).

**Comment by bastibe** [comment]
Thank you for your deep analysis!

Org-static-blog decides who gets re-rendered in the function [`org-static-blog-needs-publishing-p`](https://github.com/bastibe/org-static-blog/blob/master/org-static-blog.el#L403), which simply checks file modification dates of `pub-filename` and `post-filename`. I bet it trips up with the `pub-filename` generation, probably due to some unresolved path thing. Emacs and path normalization is often an issue.

But I have no idea how that could be related to having the config in a different place.

Regarding dates, the `#+date` (and `#+title`) field is extracted from the post with a simple regex search. It would be simple to additionally extract some other fields. It's not quite so simple to include them in your templates, however, as the formatting is currently implemented with the builtin `format-time-string`, which uses a single date.
